examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts

```diff
@@ -86,100 +108,133 @@ export class Migrator extends TypedEventEmitter<IMigratorEvents> implements IMig
+                this._preparedDetachedModel = detachedModel;
+            };
 
-            // Again, it could be the case that someone else finished the migration during our attach.
-            // Casting to MigrationState because TS doesn't understand that the state may have changed during the
-            // above await.
-            if (migratable.migrationTool.migrationState as MigrationState === "migrated") {
-                this._migrationP = undefined;
-                migratedModel.close();
-                this.takeAppropriateActionForCurrentMigratable();
-                return;
-            }
+            const completeTheMigration = async () => {
+                assert(this._preparedDetachedModel !== undefined, "this._preparedDetachedModel should be defined");
 
-            // TODO: Support retry
-            await migratable.migrationTool.finalizeMigration(containerId);
+                // Volunteer to complete the migration.
+                let isAssigned: boolean;
+                try {
+                    isAssigned = await this.currentModel.migrationTool.volunteerForMigration();
+                } catch (error) {
+                    // volunteerForMigration() will throw an error on disconnection. In this case, we should exit and
+                    // re-enter the state machine which will wait until we reconnect.
+                    // Note: while we wait to reconnect it is possible that another client will have already completed
+                    // the migration.
+                    assert(!this.connected, "We should be disconnected");
+                    return;
+                }
 
-            // If someone else finalized the migration before us, we should close the one we created.
-            if (migratable.migrationTool.newContainerId !== containerId) {
-                migratedModel.close();
-            }
+                if (!isAssigned) {
+                    // If we are not assigned, then the migration was completed by another client and we can exit.
+                    return;
+                }
 
-            // Note that we do not assume the migratedModel is the correct new one, and let it fall out of scope
-            // intentionally.  This is because if we don't win the race to set the container, it will be the wrong
-            // container/model to use.  There could maybe be some efficiency gain by retaining the model in the
-            // case that we win the race?  But it likely just doesn't matter that much because the Loader probably
-            // cached the Container anyway.
+                if (this._containerId === undefined) {
+                    this._containerId = await this._preparedDetachedModel.attach();
+                }
 
-            this._migrationP = undefined;
+                // Check to make sure we still have the task assignment.
+                if (!this.currentModel.migrationTool.haveMigrationTask()) {
```

Comment:
Would you have to do this after every async call, while you have lock? As we chatted it may be helpful to have some taskmanager helper function that helps you manage the flow.

Reply:
That is correct. I decided to keep the check here since this is a relatively small task. If the task was larger with more async calls, I would probably opt to take a different approach that relies on listening to the `"lost"` event on TaskManager to determine if we need to exit the flow.

Reply:
OK, I see. That makes sense. I would say, these examples often become reference usage for various concepts. Would promise race between "lost" event and async task be considered good simple general pattern? If so, would that be appropriate solution to incorporate here? Given this is simple use case, I'll leave it up to you. Feel free to resolve the conversation with whichever direction you conclude on.